how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb

```diff
@@ -0,0 +1,400 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for (i, lr) in enumerate(learning_rates):\n",
+    "    # Log hyperparameters\n",
+    "    run[f\"trials/{i}/parms\"] = parameters\n",
+    "    run[f\"trials/{i}/parms/lr\"] = lr\n",
+    "\n",
+    "    optimizer = optim.SGD(model.parameters(), lr=lr)\n",
+    "    for _ in trange(parameters[\"epochs\"]):\n",
+    "        for (x, y) in trainloader:\n",
+    "\n",
+    "            x, y = x.to(parameters[\"device\"]), y.to(parameters[\"device\"])\n",
+    "            optimizer.zero_grad()\n",
+    "            outputs = model.forward(x)\n",
+    "            loss = criterion(outputs, y)\n",
+    "\n",
+    "            _, preds = torch.max(outputs, 1)\n",
+    "            acc = (torch.sum(preds == y.data)) / len(x)\n",
+    "\n",
+    "            # Log metrics\n",
+    "            run[f\"trials/{i}/training/batch/loss\"].log(loss)\n",
+    "            run[f\"trials/{i}/training/batch/acc\"].log(acc)\n",
+    "\n",
+    "            loss.backward()\n",
+    "            optimizer.step()\n",
+    "\n",
+    "# Stop logging\n",
+    "run.stop()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Log metadata from each HPO trial into separate runs\n",
+    "Create local Neptune runs to log metrics from each trial into separate run"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Training Loop"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for (i, lr) in enumerate(learning_rates):\n",
+    "    # Create a new run\n",
+    "    run = neptune.init_run(\n",
+    "        api_token=neptune.ANONYMOUS_API_TOKEN,\n",
+    "        project=\"common/pytorch-integration\",\n",
+    "        name=f\"trial-{i}\",\n",
+    "        tags=[\"trial-level\"],  # to indicate that the run only contains results from a single trial,\n",
+    "    )\n",
+    "\n",
+    "    # Log hyperparameters\n",
+    "    run[\"parms\"] = parameters\n",
+    "    run[\"parms/lr\"] = lr\n",
+    "\n",
+    "    for _ in trange(parameters[\"epochs\"]):\n",
+    "        for (x, y) in trainloader:\n",
+    "\n",
+    "            x, y = x.to(parameters[\"device\"]), y.to(parameters[\"device\"])\n",
+    "            optimizer.zero_grad()\n",
+    "            outputs = model.forward(x)\n",
+    "            loss = criterion(outputs, y)\n",
+    "\n",
+    "            _, preds = torch.max(outputs, 1)\n",
+    "            acc = (torch.sum(preds == y.data)) / len(x)\n",
+    "\n",
+    "            # Log metrics\n",
+    "            run[\"training/batch/loss\"].log(loss)\n",
+    "            run[\"training/batch/acc\"].log(acc)\n",
+    "\n",
+    "            loss.backward()\n",
+    "            optimizer.step()\n",
+    "\n",
+    "    # Wait for all the tracking calls to finish.\n",
+    "    run.wait()\n",
```

Comment:
do we need to wait given that `run.stop()` will take care of syncing?

Reply:
No, not really.

But I wanted to ensure all metadata gets logged before stopping.

Reply:
won't `run.stop()` do that before stopping the run?

Reply:
Removed it.